x-pack/plugins/apm/public/components/shared/Stacktrace/FrameHeading.tsx

```diff
@@ -35,14 +36,42 @@ const FrameHeading: React.SFC<Props> = ({ stackframe, isLibraryFrame }) => {
   const lineNumber: number = get(stackframe, 'line.number');
   return (
     <FileDetails>
-      <FileDetail>{stackframe.filename}</FileDetail> in{' '}
-      <FileDetail>{stackframe.function}</FileDetail>
-      {lineNumber > 0 && (
-        <Fragment>
-          {' at '}
-          <FileDetail>line {stackframe.line.number}</FileDetail>
-        </Fragment>
-      )}
+      <FormattedMessage
+        id="xpack.apm.stacktraceTab.exceptionLocationMessage"
+        defaultMessage="{fileName} in {functionName} {atLineNumber}"
+        values={{
+          fileName: <FileDetail>{stackframe.filename}</FileDetail>,
+          functionName: <FileDetail>{stackframe.function}</FileDetail>,
+          atLineNumber:
+            lineNumber > 0 ? (
+              <FormattedMessage
+                id="xpack.apm.stacktraceTab.exceptionLocationMessage.atLineNumberText"
+                defaultMessage="at {lineNumber}"
+                description="Part of composite text: xpack.apm.stacktraceTab.exceptionLocationMessage
+                  + xpack.apm.stacktraceTab.exceptionLocationMessage.atLineNumberText
+                  + xpack.apm.stacktraceTab.exceptionLocationMessage.lineNumberText"
+                values={{
+                  lineNumber: (
+                    <FileDetail>
+                      <FormattedMessage
+                        id="xpack.apm.stacktraceTab.exceptionLocationMessage.lineNumberText"
+                        defaultMessage="line {stackframeLineNumber}"
+                        values={{
+                          stackframeLineNumber: stackframe.line.number
+                        }}
+                        description="Part of composite text: xpack.apm.stacktraceTab.exceptionLocationMessage
+                          + xpack.apm.stacktraceTab.exceptionLocationMessage.atLineNumberText
+                          + xpack.apm.stacktraceTab.exceptionLocationMessage.lineNumberText"
+                      />
+                    </FileDetail>
+                  )
+                }}
+              />
+            ) : (
+              ''
+            )
+        }}
+      />
```

Comment:
I think this is super hard to read. What do you think about this?

```ts
export function FrameHeading({ stackframe, isLibraryFrame }: Props) {
  const FileDetail = isLibraryFrame
    ? LibraryFrameFileDetail
    : AppFrameFileDetail;
  const lineNumber: number = get(stackframe, 'line.number');

  const messageWithLineNumber = () => (
    <FormattedMessage
      id="xpack.apm.stacktraceTab.exceptionLocationMessageWithLineNumber"
      defaultMessage="{fileName} in {functionName} at {lineNumber}"
      values={{
        fileName: <FileDetail>{stackframe.filename}</FileDetail>,
        functionName: <FileDetail>{stackframe.function}</FileDetail>,
        lineNumber: <FileDetail>line {lineNumber}</FileDetail>
      }}
    />
  );

  const messageWithOutLineNumber = () => (
    <FormattedMessage
      id="xpack.apm.stacktraceTab.exceptionLocationMessageWithoutLineNumber"
      defaultMessage="{fileName} in {functionName}"
      values={{
        fileName: <FileDetail>{stackframe.filename}</FileDetail>,
        functionName: <FileDetail>{stackframe.function}</FileDetail>
      }}
    />
  );

  return (
    <FileDetails>
      {lineNumber > 0 ? messageWithLineNumber() : messageWithOutLineNumber()}
    </FileDetails>
  );
}
```

Reply:
Yes, it's nicer. Thank you.